usr/lib/common/asn1.c

```diff
@@ -3473,3 +3574,806 @@ CK_RV ber_decode_DHPublicKey(CK_BYTE *data,
+    }
+
+    /* Decode sequence:
+     *     SEQUENCE (2 elem)
+     *       BIT STRING = rho
+     *       BIT STRING = t1
+     */
+    rc = ber_decode_SEQUENCE(val, &seq, &seq_len, &field_len);
+    if (rc != CKR_OK) {
+        TRACE_DEVEL("ber_decode_SEQUENCE failed\n");
+        return rc;
+    }
+
+    /* Decode rho */
+    rc = ber_decode_BIT_STRING(seq, &rho, &rho_len, &field_len);
+    if (rc != CKR_OK) {
+        TRACE_DEVEL("ber_decode_INTEGER failed\n");
+        return rc;
+    }
+
+    /* Decode t1 */
+    offset = field_len;
+    rc = ber_decode_BIT_STRING(seq + offset, &t1, &t1_len, &field_len);
+    if (rc != CKR_OK) {
+        TRACE_DEVEL("ber_decode_INTEGER failed\n");
+        return rc;
+    }
+
+    /* Build rho attribute */
+    rc = build_attribute(CKA_IBM_DILITHIUM_RHO, rho, rho_len, &rho_attr_temp);
+    if (rc != CKR_OK) {
+        TRACE_DEVEL("build_attribute failed\n");
+        goto cleanup;
+    }
+
+    /* Build t1 attribute */
+    rc = build_attribute(CKA_IBM_DILITHIUM_T1, t1, t1_len, &t1_attr_temp);
+    if (rc != CKR_OK) {
+        TRACE_DEVEL("build_attribute failed\n");
+        goto cleanup;
+    }
+
+    *rho_attr = rho_attr_temp;
+    *t1_attr = t1_attr_temp;
+
+    return CKR_OK;
+
+cleanup:
+    if (rho_attr_temp)
+        free(rho_attr_temp);
+    if (t1_attr_temp)
+        free(t1_attr_temp);
+
+    return rc;
+}
+
+/**
+ * An IBM Dilithium private key is given by:
+ *
+ *     DilithiumPrivateKey ::= SEQUENCE {
+ *       version INTEGER,     -- v0, reserved 0
+ *       rho BIT STRING,      -- nonce
+ *       key BIT STRING,      -- key/seed/D
+ *       tr  BIT STRING,      -- PRF bytes ('CRH' in spec)
+ *       s1  BIT STRING,      -- vector(L)
+ *       s2  BIT STRING,      -- vector(K)
+ *       t0  BIT STRING       -- low bits(vector L)
+ *       t1 [0] IMPLICIT OPTIONAL {
+ *         t1  BIT STRING     -- high bits(vector L)  -- see also public key
+ *       }
+ *     }
+ */
+CK_RV ber_encode_IBM_DilithiumPrivateKey(CK_BBOOL length_only,
+                               CK_BYTE **data,
+                               CK_ULONG *data_len,
+                               CK_ATTRIBUTE *rho,
+                               CK_ATTRIBUTE *seed,
+                               CK_ATTRIBUTE *tr,
+                               CK_ATTRIBUTE *s1,
+                               CK_ATTRIBUTE *s2,
+                               CK_ATTRIBUTE *t0,
+                               CK_ATTRIBUTE *t1,
+                               CK_ATTRIBUTE *opaque)
+{
+    CK_BYTE *buf = NULL;
+    CK_BYTE *buf2 = NULL;
+    CK_ULONG len, offset;
+    CK_BYTE version[] = { 0 };
+    CK_RV rc;
+    BerElement *ber;
+    BerValue *val;
+
+    /* Calculate storage for sequence */
+    offset = 0;
+    rc = 0;
+
+    rc |= ber_encode_INTEGER(TRUE, NULL, &len, NULL, sizeof(version));
+    offset += len;
+    if (opaque != NULL) {
+        rc |= ber_encode_BIT_STRING(TRUE, NULL, &len, NULL, opaque->ulValueLen, 0);
```

Comment:
Shouldn't this be ber_encode_OCTET_STRING here? Below you encode it as an octet string

Reply:
Fixed.